enumerable.rb

```diff
@@ -0,0 +1,155 @@
+    # evaluates all elements in array and return true if all of them meets the given condition
+    return test_param_all(param) unless param.nil?
+
+    all_true = true
+    if block_given?
+      my_each { |x| all_true = false unless yield(x) }
+    else
+      my_each { |x| all_true = false if x.nil? || x == false }
+    end
+    all_true
+  end
+
+  def test_param_all(param)
+    if param.class == Regexp
+      my_all? { |x| x =~ param }
+    elsif param.class == Class
+      my_all? { |x| x.is_a? param }
+    else
+      my_all? { |x| x == param }
+    end
+  end
+
+  def my_any?(param = nil)
+    # evaluates all elements in array and return true if any of them meets the given condition
+    return test_param_any(param) unless param.nil?
+
+    any_true = false
+    if block_given?
+      my_each { |x| any_true = true if yield(x) }
+    else
+      my_each { |x| any_true = true if !x.nil? && x != false }
+    end
+    any_true
+  end
+
+  def test_param_any(param)
+    if param.class == Regexp
+      my_any? { |x| x =~ param }
+    elsif param.class == Class
+      my_any? { |x| x.is_a? param }
+    else
+      my_any? { |x| x == param }
+    end
+  end
+
+  def my_none?(param = nil)
+    # evaluates all elements in array and return true if all of them are false
+    return test_param_none(param) unless param.nil?
+
+    all_false = true
+    if block_given?
+      my_each { |x| all_false = false if yield(x) }
+    else
+      my_each { |x| all_false = false unless x.nil? || x == false }
+    end
+    all_false
+  end
+
+  def test_param_none(param)
+    if param.class == Regexp
+      my_none? { |x| x =~ param }
+    elsif param.class == Class
+      my_none? { |x| x.is_a? param }
+    else
+      my_none? { |x| x == param }
+    end
+  end
+
+  def my_count(param = nil)
+    # count the number of items that meets the condition given
+    total = 0
+    if block_given?
+      my_each { |x| total += 1 if yield(x) }
+    elsif !param.nil?
+      my_each { |x| total += 1 if x == param }
+    else
+      return length
+    end
+    total
+  end
+
+  def my_map(proc = nil)
+    return to_enum :my_map unless block_given? || proc.class == Proc
+
+    new_array = []
+    if proc.class == Proc
+      my_each { |x| new_array << proc.call(x) }
+    elsif block_given?
+      my_each { |x| new_array << yield(x) }
+    end
+    new_array
+  end
+
+  def my_inject(init = nil, sym = nil)
+    total = self[0]
+    i = 1
+    if init.class == Symbol
+      while i < length
+        total = total.send(init, self[i])
+        i += 1
+      end
+    elsif !init.nil? && init.class == Symbol
+      total = init
+      my_each { |n| total = total.send(sym, n) }
+    elsif init.nil? && sym.nil?
+      while i < length
+        total = yield(total, self[i]) if block_given?
+        i += 1
+      end
+    elsif !init.nil? && sym.nil?
+      total = init
+      my_each { |n| total = yield(total, n) } if block_given?
+    end
+    total
+  end
```

Comment:
When testing with a range it gives this error: undefined method `[]' for ... Range

- [ ] Let's work on my_inject so it also works with ranges:

> this (5..10).my_inject(2) { |product, n| product * n } should be equal to this (5..10).inject(2) { |product, n| product * n }

- [ ] Let's handle cases where a block is not given and a Symbol or String operator is passed instead.

For example: [1,2,3].my_inject(:+) should equal [1,2,3].inject(:+)

HINT: Investigate the [send](https://ruby-doc.org/core-2.6.3/Object.html#method-i-send) method. Since 2.send(:+, 2) == 4

Reply:
'my_inject' methods fixed to take ranges and Symbols.